dask/dataframe/shuffle.py

```diff
@@ -24,6 +25,196 @@
 logger = logging.getLogger(__name__)
 
 
+class ShuffleStage(Layer):
+    """Shuffle HighLevelGraph Layer
+
+    High-level graph layer corresponding to a single stage of
+    an inter-partition shuffle operation.
+
+    Stage: (shuffle-group) -> (shuffle-split) -> (shuffle-join)
+
+    Parameters
+    ----------
+    name : str
+        Name of new (partially) shuffled collection.
+    column : str or list of str
+        Column(s) to be used to map rows to output partitions (by hashing).
+    inputs : list of tuples
+        Each tuple dictates the data movement for a specific partition.
+    stage : int
+        Index of the current shuffle stage.
+    npartitions : int
+        Number of output partitions for the full (multi-stage) shuffle.
+    n : int
+        Number of partitions in the original (un-shuffled) DataFrame.
+    k : int
+        A partition is split into this many groups during each stage.
+    ignore_index: bool, default False
+        Ignore index during shuffle.  If ``True``, performance may improve,
+        but index values will not be preserved.
+    dep_name : str
+        Name of input collection.
+    dep_meta : pd.DataFrame-like object
+        Empty metadata of input collection.
+    parts_out : list of int (optional)
+        List of required output-partition indices.
+    """
+
+    def __init__(
+        self,
+        name,
+        column,
+        inputs,
+        stage,
+        npartitions,
+        n,
+        k,
+        ignore_index,
+        dep_name,
+        dep_meta,
+        parts_out=None,
+    ):
+        self.column = column
+        self.name = name
+        self.inputs = inputs
+        self.stage = stage
+        self.npartitions = npartitions
+        self.n = n
+        self.k = k
+        self.ignore_index = ignore_index
+        self._name = dep_name
+        self._meta = dep_meta
+        self.parts_out = parts_out or range(len(inputs))
+
+    def __repr__(self):
+        return "ShuffleStage<name='{}', stage={}, k={}, npartitions={}>".format(
+            self.name, self.stage, self.k, self.npartitions
+        )
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict["dsk"]
```

Comment:
What's the reason for `_cached_dict`nesting the Dask dict under the `"dsk"` key, rather than just returning the `dsk` itself?

Reply:
No reason at all - Another detail I forgot to clean up.  The very first version of the shuffle layer was doing something similar to Blockwise, and so we were caching a BasicLayer object as well as `dsk`.